api/controllers/AuthController.js

```diff
@@ -25,19 +25,33 @@ module.exports = {
      * might fail to complete the login and then be confused when they get redirected somewhere unexpected the next time they visit the site. */
     var login_info = {type: req.query.loginType, redirect: req.query.redirect || '/', validation: crypto.randomBytes(32).toString('hex')};
     req.session.validation = login_info.validation;
-    var auth_data = {state: JSON.stringify(login_info), duration: 'permanent', failureRedirect: '/login'};
-    if (req.query.loginType !== 'mod') {
-      auth_data.scope = 'identity';
+    var auth_data = {state: JSON.stringify(login_info)};
+    if (req.query.loginType === 'mod') {
+      auth_data.duration = 'permanent'; //Mods have a permanent access token. Scope is not specified here, so the scope from config/express.js is used.
+    } else {
+      auth_data.scope = 'identity'; //Regular users only need a temporary access token with reduced scope.
     }
     passport.authenticate('reddit', auth_data)(req, res);
   },
 
   callback: function (req, res) {
     passport.authenticate('reddit', {duration: 'permanent', failureRedirect: '/login'}, async function (err, user) {
-      let login_info = JSON.parse(req.query.state);
+      if (err) {
+        if (err === 'banned') {
+          return res.view(403, {error: 'You have been banned from FlairHQ'});
+        }
+        return res.view(403, {error: 'Sorry, something went wrong. Try logging in again.'});
+      }
+      var login_info;
+      try {
+        login_info = JSON.parse(req.query.state);
+      } catch (err) {
+        console.log('Error with parsing /u/' + user.name + '\'s session state');
+        return res.serverError(err);
+      }
       if (login_info.validation !== req.session.validation) {
-        sails.warn("Failed login for /u/" + user.name + ": invalid session state");
-        return res.forbidden();
+        console.log("Failed login for /u/" + user.name + ": invalid session state");
+        return res.view(403, {error: 'You have an invalid session state. (Try logging in again.)'});
```

Comment:
Just refactored some stuff here. Also, people will be told when they have an invalid session state or are banned, rather than just seeing the default 403 page.